### PR TITLE
health check: split and move health checker impl

### PR DESCRIPTION
**Pull request by dio**
health check: split and move health checker impl

This patch splits the health checker impl into two folds: 

1. Base implementation 
2. The implementation of TCP, HTTP and gRPC health checkers. 

```
source/
`-- common/
    `-- upstream/
        |-- BUILD
        |-- ...
        |-- health_checker_base_impl.cc
        |-- health_checker_base_impl.h
        |-- health_checker_impl.cc
        |-- health_checker_impl.h
        |-- ...
```

The reason for splitting is to let the health checker base to be extended by the
upcoming (pluggable) extension health checkers https://github.com/envoyproxy/envoy/issues/2933.

*Risk Level*: 
Low

*Testing*:
No new tests are introduced.

*Docs Changes*:
N/A

*Release Notes*:
N/A

Partially, this is an effort to deal with: 
- https://github.com/envoyproxy/envoy/issues/1276

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

### CODE REVIEW

File: source/common/upstream/health_checker/health_checker_base.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
+#include "envoy/api/v2/core/health_check.pb.h"
+#include "envoy/event/timer.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/upstream/health_checker.h"
+
+#include "common/common/logger.h"
+
+namespace Envoy {
+namespace Upstream {
+
```
Comment (dio) [inline source/common/upstream/health_checker/health_checker_base.h]: Not sure if we want to introduce new namespace here (probably `HealthCheckers` and the dir name will be changed to `health_checkers`? Please help with some suggestions).

**Comment by mattklein123** [comment]
@dio what's the rationale behind the directory moves? I think it's fine if we want to do that, but it's not explicitly necessary (could just split the file apart in the existing directory)? (I'm fine either way just want to understand the plan).

**Comment by dio** [comment]
@mattklein123 nothing is special. I glanced over the `source/common/upstream` dir, I was under impression that each topic has only a pair of `*.{h, cc}` (yes, until I saw the `cds_api` and `cds_subscription`). Also the treatment with lua, but probably that is different. 

I'm overthinking it. Will just put it in the upstream directory then.

One more question: Do you think `health_checker_base_impl.{h, cc}` and `health_checker_impl.{h, cc}` are fine? (thinking the file names are too long).

**Comment by mattklein123** [comment]
> One more question: Do you think health_checker_base_impl.{h, cc} and health_checker_impl.{h, cc} are fine? (thinking the file names are too long).

It's fine IMO. I think moving the files into their own directory/namespace is also fine, but probably not worth the review churn right now.

**Pull request merged**